### PR TITLE
feat: initial support for improved Release definition

### LINKED ISSUE #14: List all artifacts for a given product in request template

### LINKED ISSUE #15: Extract artifacts for a given product based on BOM

### PR DESCRIPTION
**Pull request by metacosm**
Fixes #15 
Fixes #14

### CODE REVIEW

File: src/main/java/dev/snowdrop/jira/atlassian/POMParser.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * @author <a href="claprun@redhat.com">Christophe Laprun</a>
+ */
+public class POMParser {
+	public static List<Artifact> getArtifactsWith(String property) {
+		try {
+			MavenXpp3Reader mavenreader = new MavenXpp3Reader();
+			final URL resource = POMParser.class.getClassLoader().getResource("pom.xml");
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/POMParser.java:33]: The `pom.xml` file to be parsed should not be hard coded but instead be an URL pointing to a maven upstream project released on maven central such as : Spring Boot Dependencies, Spring cloud ...

File: src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java [hunk 2]
```diff
@@ -79,13 +79,12 @@ public static void createComponentIssues() {
 		Release release = (Release) pojo;
 
 		for (Component component : release.getComponents()) {
-			if (!component.getSkipCreation()) {
-				IssueInputBuilder iib = new IssueInputBuilder();
-				iib.setProjectKey(component.getJiraProject());
-				iib.setSummary(component.getJiraTitle());
-				iib.setDescription(generateIssueDescription(release, component));
-				iib.setIssueType(TASK_TYPE());
-				iib.setDueDate(toDateTime(release.getDueDate()));
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java:88]: For minor releases, we will not contact the JIRA stakeholders as we will reuse the existing component version or the version as defined within the CVEs. So, we still need this option that you deleted

File: src/main/java/dev/snowdrop/jira/atlassian/model/Release.java [hunk 3]
```diff
@@ -1,50 +1,28 @@
 package dev.snowdrop.jira.atlassian.model;
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
 import java.util.List;
 
 public class Release {
-	private String jiraProject;
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/model/Release.java:6]: How will you been able to create a JIRA issue within the target JIRA project without this field ?

File: src/main/java/dev/snowdrop/jira/atlassian/model/Release.java [hunk 4]
```diff
@@ -1,50 +1,28 @@
...
-	private String jiraKey;
-	private String version;
-	private String longVersionName;
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/model/Release.java:9]: We still need this long title description as it is used to create the JIRA component issue

File: src/main/java/dev/snowdrop/jira/atlassian/model/Release.java [hunk 5]
```diff
@@ -1,50 +1,28 @@
...
-	private String fixVersion;
-	private String EOL;
-	private String releaseDate;
-	private String dueDate;
-	private String dueDateFormatted;
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/model/Release.java:14]: Such fields are needed to fill the mustache template with dueDate, EOL & our release date

**Comment by cmoulliard** [comment]
FYI: The `mustache.description` needs to be reviewed too ;-)

File: src/main/java/dev/snowdrop/jira/atlassian/model/Component.java [hunk 6]
```diff
@@ -1,62 +1,35 @@
...
-    }
+public class Component {
+    @JsonProperty
+    private Issue issue;
 
-    public void setName(String name) {
-        this.name = name;
-    }
+    @JsonProperty
+    private List<String> properties;
```
Comment (cmoulliard) [inline src/main/java/dev/snowdrop/jira/atlassian/model/Component.java:15]: What is a property ? A component/artifact ?

File: src/test/java/dev/snowdrop/jira/ValidateYAMLtoPojoTest.java [hunk 7]
```diff
@@ -3,13 +3,16 @@
...
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
 import java.io.File;
 import java.io.IOException;
 import java.net.URL;
+import java.util.List;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
```
Comment (cmoulliard) [inline src/test/java/dev/snowdrop/jira/ValidateYAMLtoPojoTest.java:18]: This test class should include as that was the case before `@Before` loading to the mustache template to validate the description generated from the pojo

File: src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java (hunk 2 above)
Comment (metacosm) [reply src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java:88]: No. Stakeholders should always be contacted. Every release we put out should be vetted by stakeholders, minor release or not.

**Comment by metacosm** [comment]
For an example of the result of this PR, see https://issues.redhat.com/browse/SB-1503, which was updated automatically using the following command: ` java -jar ./target/uber-issues-manager-1.0.0-SNAPSHOT.jar -user <user> -password <password> -action create-component -git "metacosm/spring-boot-bom/release-integration"`. 

The `git` parameter defines the repository reference from which all the information is retrieved: `release.yml` and associated POM information. 

You can see the repository layout here: https://github.com/metacosm/spring-boot-bom/tree/release-integration

File: src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java (hunk 2 above)
Comment (cmoulliard) [reply src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java:88]: @zekaguilar Do you also agree that we contact the stakeholders (= create JIRA tickets within their project) for minor releases even if we know that we dont expect from them a new release of their component but instead a QE/Test validation that their component or starter will work with by example Spring Boot 2.x.y ?

Comment (metacosm) [reply src/main/java/dev/snowdrop/jira/atlassian/ReleaseService.java:88]: We never expect a new release. We don't care about new releases, we just want a *supported* release for our Spring Boot version for the duration of the release life time. That's why we always need to get this information from stakeholders.

**Comment by cmoulliard** [comment]
> The `git` parameter defines the repository reference from which all the information is retrieved: `release.yml` and associated POM information.

If we use our Spring Boot POM.xml, that means implicitly that someone from the team will update it manually prior to do anything else.  My idea was instead to fetch from the different Spring Boot, Spring cloud , ... POM/BOM already released upstream the version to inform the stakeholders about the targeted version (they could potentially release). WDYT : @metacosm @zekaguilar

**Comment by metacosm** [comment]
> > The `git` parameter defines the repository reference from which all the information is retrieved: `release.yml` and associated POM information.
> 
> If we use our Spring Boot POM.xml, that means implicitly that someone from the team will update it manually prior to do anything else. My idea was instead to fetch from the different Spring Boot, Spring cloud , ... POM/BOM already released upstream the version to inform the stakeholders about the targeted version (they could potentially release). WDYT : @metacosm @zekaguilar

Have you checked the generated issue and looked at the `release.yaml`? Unless I'm mistaken the code already does what you're suggesting.

**Comment by metacosm** [comment]
Also note that there's no manual update of our upstream POM: this is done by our release process…

**Comment by cmoulliard** [comment]
> Also note that there's no manual update of our upstream POM: this is done by our release process…

Can you then explain how we `got/calculate` then for Hibernate & Hibernate validator component such versions ?
```
org.hibernate:hibernate-core (upstream version is 5.4.14.Final)
org.hibernate:hibernate-entitymanager (upstream version is 5.4.14.Final)
org.hibernate.validator:hibernate-validator (upstream version is 6.0.18.Final)
```

**Comment by metacosm** [comment]
It's extracted from our Pom which is automatically updated from upstream with atomist

**Comment by cmoulliard** [comment]
2 minor changes are needed for the mustache template: 

1) Missing data - EOL

The following part of the generated text shows `**` instead of the `*EOL*`
```
We will support this new release until ** and therefore require compatibility and support
statement from your team for this Spring Boot version in the form of the following information:
```

2) Non needed CR

The following text could be improved to avoid to add non needed CRs

```
Not that we don't require a new release of your product as long as the artifacts you provide us with are tested and
supported on Spring Boot 2.3.1.RELEASE
until 30 Sep 2021.
```
-->
```
Not that we don't require a new release of your product as long as the artifacts you provide us with are tested and supported on Spring Boot 2.3.1.RELEASE until 30 Sep 2021.
```

**Comment by metacosm** [comment]
Ideally, the component request would occur before the Pom update from atomist but there is no point in duplicating what atomist already does and which is needed to be performed at some point early in the release process anyway.

**Comment by cmoulliard** [comment]
> It's extracted from our Pom which is automatically updated from upstream with atomist

Fantastic as this first step of the process is fully automated and GAVs calculated from upstream Spring BOM :-)

**Comment by metacosm** [comment]
> 2 minor changes are needed for the mustache template:
> 
>     1. Missing data - EOL
> The following part of the generated text shows `**` instead of the `*EOL*`

This is already fixed in the code, I just didn't re-generate the issues after I had fixed this issue. 

> The following text could be improved to avoid to add non needed CRs
> 
> ```
> Not that we don't require a new release of your product as long as the artifacts you provide us with are tested and
> supported on Spring Boot 2.3.1.RELEASE
> until 30 Sep 2021.
> ```
> 
> -->
> 
> ```
> Not that we don't require a new release of your product as long as the artifacts you provide us with are tested and supported on Spring Boot 2.3.1.RELEASE until 30 Sep 2021.
> ```

I don't understand what you want me to change here.

**Comment by metacosm** [comment]
Can we merge this, @cmoulliard?

**Comment by cmoulliard** [comment]
> Can we merge this, @cmoulliard?

Did you fixed the Carriage return issue - see remark ? @metacosm

**Pull request merged**